enterprise/backend/src/metabase_enterprise/transforms/api.clj

```diff
@@ -281,6 +275,86 @@
                               :run_id run-id})
           (assoc :status 202)))))
 
+(defn- extract-columns-from-query
+  "Attempts to extract column names from an MBQL query without executing it.
+
+  Returns a vector of column names (as strings), or nil if extraction fails.
+
+  The query is first compiled to native SQL, hen uses PreparedStatement.getMetaData()
+  to inspect the query structure. This works for most modern JDBC drivers but may not
+  be supported by all drivers or for all query types."
+  [driver database-id query]
+  (try
+    (let [{:keys [query]} (qp.compile/compile query)]
+      (sql-jdbc.execute/do-with-connection-with-options
+       driver
+       database-id
+       {}
+       (fn [conn]
+         (with-open [^PreparedStatement stmt (sql-jdbc.execute/prepared-statement driver conn query [])]
+           (when-let [rsmeta (.getMetaData stmt)]
+             (let [columns (sql-jdbc.execute/column-metadata driver rsmeta)]
+               (seq (map :name columns))))))))
```

Comment:
I would personally hesitate at letting a seq escape with handles to open stmt derived stuff, but have not investigated lifetime guarantees or lack of here in JDBC